packages/gestalt/src/FormHelperText.js

```diff
@@ -1,21 +1,92 @@
 // @flow strict
-
-import { type Node } from 'react';
+import { Fragment, useRef, useEffect, useState, type Node } from 'react';
 import Box from './Box.js';
+import Flex from './Flex.js';
+import Status from './Status.js';
 import Text from './Text.js';
+import { type MaxLength } from './TextField.js';
 
 type Props = {|
-  text: string,
+  text: ?string,
+  maxLength?: ?MaxLength,
+  currentLength?: number,
   addA11yPause?: boolean,
 |};
 
-export default function FormHelperText({ text, addA11yPause = false }: Props): Node {
+export default function FormHelperText({
+  currentLength,
+  text,
+  maxLength,
+  addA11yPause = false,
+}: Props): Node {
+  const ref = useRef();
+  const [width, setWidth] = useState(undefined);
+
+  useEffect(() => {
+    const containerWidth = ref?.current?.getBoundingClientRect().width;
+    setWidth(containerWidth ? Math.ceil(containerWidth) : undefined);
+  }, [ref]);
+
+  const maxLengthChars = maxLength?.characterCount.toString() ?? '';
+
+  const maxLengthReached = (currentLength ?? 0) >= (maxLength?.characterCount ?? 0);
+
+  let status = 'warning';
+  let textColor = 'warning';
+
+  if (
+    typeof currentLength === 'number' &&
+    typeof maxLength?.characterCount === 'number' &&
+    currentLength > maxLength?.characterCount
+  ) {
+    status = 'problem';
+    textColor = 'error';
+  }
+
   return (
     <Box marginTop={2}>
-      <Text color="subtle" size="100">
-        {addA11yPause ? <Box display="visuallyHidden">:</Box> : null}
-        {text}
-      </Text>
+      <Flex gap={4}>
+        <Flex.Item flex="grow">
+          {text ? (
+            <Text color="subtle" size="100">
+              {addA11yPause ? <Box display="visuallyHidden">:</Box> : null}
+              {text}
+            </Text>
+          ) : null}
+        </Flex.Item>
+        {maxLength ? (
```

Comment:
Consider pulling this out to a separate component, which will also encapsulate all the logic from lines 30-44

Reply:
DONE